stateresolutionv2.go

```diff
@@ -0,0 +1,483 @@
+			if event.StateKey() == nil || *event.StateKey() == "" {
+				r.resolvedPowerLevels = &event
+			}
+		case MRoomJoinRules:
+			// Join rule events are only valid with an empty state key.
+			if event.StateKey() == nil || *event.StateKey() == "" {
+				r.resolvedJoinRules = &event
+			}
+		case MRoomThirdPartyInvite:
+			// Third party invite events are only valid with a non-empty state key.
+			if event.StateKey() != nil && *event.StateKey() != "" {
+				r.resolvedThirdPartyInvites[*event.StateKey()] = &event
+			}
+		case MRoomMember:
+			// Membership events are only valid with a non-empty state key.
+			if event.StateKey() != nil && *event.StateKey() != "" {
+				r.resolvedMembers[*event.StateKey()] = &event
+			}
+		}
+	}
+}
+
+// eventMapFromEvents takes a list of events and returns a map, where the key
+// for each value is the event ID.
+func eventMapFromEvents(events []Event) map[string]Event {
+	r := make(map[string]Event)
+	for _, e := range events {
+		r[e.EventID()] = e
+	}
+	return r
+}
+
+// separate takes a list of events and works out which events are conflicted and
+// which are unconflicted.
+func separate(events []Event) (conflicted, unconflicted []Event) {
+	// The stack maps event type -> event state key -> list of state events.
+	stack := make(map[string]map[string][]Event)
+	// Prepare the map.
+	for _, event := range events {
+		// If we haven't encountered an entry of this type yet, create an entry.
+		if _, ok := stack[event.Type()]; !ok {
+			stack[event.Type()] = make(map[string][]Event)
+		}
+		// Add the event to the map.
+		stack[event.Type()][*event.StateKey()] = append(
+			stack[event.Type()][*event.StateKey()], event,
+		)
+	}
+	// Now we need to work out which of these events are conflicted. An event is
+	// conflicted if there is more than one entry for the (type, statekey) tuple.
+	// If we encounter these events, add them to their relevant conflicted list.
+	for _, eventsOfType := range stack {
+		for _, eventsOfStateKey := range eventsOfType {
+			if len(eventsOfStateKey) > 1 {
+				// We have more than one event for the (type, statekey) tuple, therefore
+				// these are conflicted.
+				conflicted = append(conflicted, eventsOfStateKey...)
+			} else if len(eventsOfStateKey) == 1 {
+				unconflicted = append(unconflicted, eventsOfStateKey[0])
+			}
+		}
+	}
+	return
+}
+
+// prepareConflictedEvents takes the input power level events and wraps them in
+// stateResV2ConflictedPowerLevel structs so that we have the necessary
+// information pre-calculated ahead of sorting.
+func (r *stateResolverV2) prepareConflictedEvents(events []Event) []stateResV2ConflictedPowerLevel {
+	block := make([]stateResV2ConflictedPowerLevel, len(events))
+	for i, event := range events {
+		block[i] = stateResV2ConflictedPowerLevel{
+			powerLevel:     r.getPowerLevelFromAuthEvents(event),
+			originServerTS: int64(event.OriginServerTS()),
+			eventID:        event.EventID(),
+			event:          event,
```

Comment:
We don't update them, but it's mostly for two reasons:

1. It stops us from having to repeat the same function calls, in case they were doing (or may end up doing in the future) something more expensive
1. It means that we have exactly the details we need for sorting in one struct

Reply:
I like your reasoning, but then we should be doing that everywhere for that struct so we don't need to store `Event` at all. We currently call `event.AuthEventIDs()` so can we make that a field and remove `event` entirely please?